crates/sui-core/src/transaction_input_checker.rs

```diff
@@ -192,6 +197,58 @@ async fn check_gas(
     }
 }
 
+/// Run the bytecode verifier with a timeout of `timeout_us`
+/// This function only fails if the verification does not complete within the timeout
+/// If the actual verification succeeds or fails within the timeout, the function still succeeds
+#[instrument(level = "trace", skip_all)]
+async fn run_timed_bytecode_verifier(
+    transaction: &TransactionData,
+    timeout_us: u64,
+) -> Result<(), SuiError> {
+    async fn verify_package_inner(
+        move_module_publish: &MoveModulePublish,
+        timeout_us: u64,
+    ) -> Result<(), SuiError> {
+        let mods = move_module_publish.clone();
+
+        let task = async move {
+            let modules = make_compiled_modules(&mods.modules).map_err(|e| {
+                SuiError::ModuleDeserializationFailure {
+                    error: e.to_string(),
+                }
+            })?;
+
+            let fut = async move {
+                // We don't actually care if it passes or fails. We just care that it completes on time.
+                let _ = sui_adapter::adapter::run_move_and_sui_bytecode_verifier(&modules);
+            };
+
+            select! {
+                () = sleep(Duration::from_micros(timeout_us))=> {
+                    Ok::<bool, SuiError>(false)
+                }
+                () = fut => {
+                    Ok(true)
+                }
+            }
+        };
+        if !task.await? {
+            return Err(SuiError::ModuleVerificationFailure {
+                error: format!("Move package verification exceeded timeout {timeout_us}us"),
+            });
+        }
+
+        Ok(())
+    }
+    for tx in transaction.kind.single_transactions() {
```

Comment:
Do we want to apply the time limit to each package or all the packages?

Reply:
this is an interesting question and with programmable transaction things can only get uglier maybe?
I would say let's not worry for now and let's leave it as is, but let's think about how this is going to work for programmable transactions